### ISSUE TITLE
Quantize network files

### ISSUE THREAD

**Issue by gcp** [gcp/leela-zero#1733]
Won't fundamentally solve #1731 but a simple and fully backwards compatible optimization:

We already know that using fp16 doesn't really affect the network strength. This means we should be able to write out the network files with much less significant digits than we are now, and maybe cut the size of everything almost in half.

Given the experience with things like Bfloat16, it may even be that less digits than in "half" format may be needed, as long as the comma is in the right spot (which it is given that we're text based).

This could be nicely tested by quantizing some existing networks and scheduling matches between them. It would be interesting to see how far we can go before the precision loss is really noticeable.

**Comment by l1t1** [comment]
write out the network files into ieee 754 float can save half space than double, but write as plain text cannot.

**Comment by gcp** [comment]
Writing the files in binary (be it half, double or Bfloat16) isn't supported by the clients nor the training code nor the server and needs a network format update. 

We already write out files with float precision, I'm saying we can use even less than that.

**Comment by ihavnoid** [comment]
Another hillarious but effective trick is to replace all instances of 0.xxx to .xxx, e.g., print .372 instead of 0.372 - most numbers are less than 1.000 so this should save quite a bit of size.

**Comment by l1t1** [comment]
compress to gz .372 is same as 0.372

[nozero.txt.gz](https://github.com/gcp/leela-zero/files/2289827/nozero.txt.gz)
[zero.txt.gz](https://github.com/gcp/leela-zero/files/2289828/zero.txt.gz)
[nozero.txt](https://github.com/gcp/leela-zero/files/2289829/nozero.txt)
[zero.txt](https://github.com/gcp/leela-zero/files/2289830/zero.txt)

date| time|size|file
-|-|-|-
2018-08-15 | 17:35            |   174 |nozero.txt
2018-08-15 | 17:36              |  60 |nozero.txt.gz
2018-08-15 | 17:35             |  203 |zero.txt
2018-08-15 | 17:36              |  59 |zero.txt.gz

**Comment by gjm11** [comment]
Leela Nozero. (Sorry.)

**Comment by gcp** [comment]
>So even under very aggressive compression settings, simply removing the 0 is a net gain.

It's still less things that have to be encoded as being present, makes sense.

**Comment by gcp** [comment]
I did a test with:

```
#!/usr/bin/env python3
import sys


def write_block(f, b):
    for w in b:
        f.write(' '.join(map(str, w.flatten())) + '\n')

with open(sys.argv[1], 'r') as w:
    with open('quant.txt', 'w') as q:
        w.readline()  # skip network version
        q.write('1\n')
        for line in w:
            weights = [float(x) for x in line.split()]
            print("weights: {}".format(len(line)))
            str_weights = " ".join([str("{:.2g}".format(x)) for x in weights])
            q.write(str_weights + "\n")
```

Which didn't do the zero stripping, round to zero etc tricks.

```
-rw-rw-r-- 1 morbo morbo 93595393 aug 15 15:10 c910dee9c057d144efaafb603ce69d564825ca316dd8129ce0f05c1231d1f9bc.gz
-rw-rw-r-- 1 morbo morbo 36229210 aug 15 15:32 quant2.txt.gz
-rw-rw-r-- 1 morbo morbo 51016744 aug 15 15:29 quant3.txt.gz
-rw-rw-r-- 1 morbo morbo 64418836 aug 15 15:27 quant4.txt.gz
-rw-rw-r-- 1 morbo morbo 79537730 aug 15 15:24 quant5.txt.gz
```

I tested quant2 (:2g) which has about 7 bits of mantissa. It's 11-11 vs the unquantized network so far.

**Comment by gcp** [comment]
I guess I should incorporate some of the tricks in this script and then do a comparison whether rounding to half style (big significand, less exponent bits, like @Ttl's script) or Bfloat16 style (small significand, keep exponent bits) is more effective for us.

**Comment by betterworld** [comment]
@Ttl 
> x = x.replace('0.', '.')

Careful with 10.1

**Comment by Ttl** [comment]
Well spotted. There actually was one weight starting with "10." in the weight file.

I calculated some L2-norms between original and quantized weights and "5f" seems to be more accurate than "5g":

```
{:.5f}, 3.53505565914826e-05
{:.3g}, 0.04279058207864573
{:.4g}, 0.00163199402697143
{:.5g}, 0.00022178429992239518
```

**Comment by gcp** [comment]
The testmatch of c910dee vs the quant2 version ended up in a 84-76 win for the quant2 version.

>"5f" seems to be more accurate than "5g"

But what about the file sizes?

**Comment by gcp** [comment]
>{:.2g} seems significantly more inaccurate than half precision.

That makes sense as it's 11 bits of significand compared to 6.5 bits. However, Bfloat only has 7 and that's apparently the hot thing in machine learning land. From the above test, it seems fine for us.

I think the question is whether you want to keep more precision in the values, or be able to handle the odd value that is very big or small? From your results, small values could just be rounded to zero. But looking in the actual network files, this means a large part of the last few innerproduct layers is fully redundant, as it's almost all tiny values.

I'm somewhat partial to using g/Bfloat16 style because it seems more robust to keep some precision in all values and not just round those to 0. But maybe that's just my preconception/imagination, as BatchNorm means very small coefficients or biases can't really have much meaning, given the distribution of input values?

Any comments/thoughts?

**Comment by Ttl** [comment]
Minimum positive non-denormal half precision number is 6e-5. Since numbers smaller than that would be in any case rounded to zero when storing them in half precision we might as well just round them already in the weight file. But I'm not completely sure if it has some drawbacks when restoring training from the quantized weights?

Just using "g" formatting should be a safe choice. Very small numbers could probably be written with less precision without much issues.

**Comment by l1t1** [comment]
my test of elf v0
```
C:\Users\aaa\Downloads>python quantize_weights.py  elf_converted_weights.txt
Weight file difference L2-norm: 0.0034445704188723683

C:\Users\aaa\Downloads>dir elf_converted_weights*

2018/05/04  02:45       260,579,331 elf_converted_weights.txt
2018/08/17  15:24       154,259,295 elf_converted_weights_quantized.txt
```
could you set a test match of elf v0 vs elf_v0_quantized?

**Comment by diadorak** [comment]
We don't need to use the ELF network. Matching a 5x64 network against the quantized version of itself should be enough, right?

**Comment by gcp** [comment]
There might be an argument that the deeper the network, the more the intermediate rounding can affect things. I'd just queue the current best vs a quantized version of itself. Or best 40b vs quant 40b.

**Comment by Mardak** [comment]
Here's some numbers from the 40b quantized test:
gzip size quantized: 96MB, original: 183MB
plain text size quantized: 352MB, original: 534MB
```
quantized:  -.00371     -.00158     -.00174     -.00292     -.00653    -.0118     -.000745     -.00946    .0194     -.0037     .000385     -.00168     -.000181   -.00304     -.0117    -.000807     -.00911
original:  -0.00370743 -0.00157953 -0.00174433 -0.00292096 -0.0065255 -0.0117573 -0.000745065 -0.0094582 0.0194095 -0.0036998 0.000384994 -0.00167523 -0.0001806 -0.00303627 -0.011729 -0.000807318 -0.0091053
```

**Comment by l1t1** [comment]
in a304 vs b072 match, two facts are changed, training games and file quantized, it is hard to judge

**Comment by l1t1** [comment]
in a304 vs b072 match, same training games and the file quantized win, it is hard to explain.

**Comment by gcp** [comment]
>it is hard to explain.

It's very easy, it's called statistical variance.

**Comment by Gondolieri** [comment]
Another idea for an explanation: Perhaps very little values should be 0.0 in a 'correct' weight file. The training didn't reach that value yet but the script erases the bad influence.
Up to now I didn't double check whether this may be true.

**Comment by gcp** [comment]
The script doesn't round to zero.

**Comment by jokkebk** [comment]
Assuming the network has not yet converged to "optimal weights" on most of the values when measured with rounded numbers, the rounding should "improve performance" about 50 % in the roundings and "reduce performance" on the other 50. 

Has anyone measured how much the values change over one passed network to another (I'm unfortunately not familiar with the weight file format so no, I haven't) -- if the unquantized change is larger on the nodes, rounding should do little. But if we have like 20 % of weights changing so little they will not get over "0.5" in one go, we'd effectively be locking the network partially into place.

Of course, if the server would have the "full precision" network for training, and it's just the training games run by clients (and maybe match games) that used the reduced version, it should be much a non-issue, seeing how close performance the quantized versions are putting out.

**Comment by gcp** [comment]
There's about 10 bits of precision in the significant, if weights don't change by a fraction of at least 1/1000 in the 256k training steps, they indeed will be locked in place. That seems fine by me!

>Of course, if the server would have the "full precision" network for training, and it's just the training games run by clients (and maybe match games) that used the reduced version, it should be much a non-issue, seeing how close performance the quantized versions are putting out.

This is partially true in that during a training run the server has the unquantized weights. But every run does restart from the actual network, which will be quantized now.

**Comment by gcp** [comment]
I changed the server so the next bunch of networks are all going to be quantized out of the box. All test results so far were extremely good. This is more or less the last step before switching to 40b, I believe.

**Comment by l1t1** [comment]
the winrate of Quantized 20b  is low

Start Date | Network Hashes | Wins / Losses | Games | SPRT
-- | -- | -- | -- | --
2018-08-22 21:49 | f37a584b  VS  7141e697 | 6 : 27 (18.18%) | 33 / 400 | fail

**Comment by l1t1** [comment]
could you test a original version of  f37a584b?

**Comment by gcp** [comment]
Looks like it was just bad luck, the other networks are more normal.

**Comment by l1t1** [comment]
68a8e58e is also bad luck.the luck of b072 is too good to believe.

**Comment by l1t1** [comment]
think about force promote 1667? it is best  Quantized 20b

**Comment by alphaladder** [comment]
@I1t1  Friend, keep some patience and trust GCP.

**Comment by gonzalezjo** [comment]
I did a lot of research on this, and wrote some code to automate tests for various approaches to shrinking network files. My work is documented here: https://github.com/LeelaChessZero/lc0/issues/121

**Comment by l1t1** [comment]
the early passed wrights is playing more games of their match

**Comment by gonzalezjo** [comment]
I collected some data on compression of leelaz-model-swa-12-96000.txt. Using an error tolerance of 1 / 2 ^ 8, the results are as follows: 



```wc -c leelaz-model-swa-12-96000.7z leelaz-model-swa-12-96000.bsc leelaz-model-swa-12-96000.uncompressed.txt leelaz-model-swa-12-96000.ppmd
14712009 leelaz-model-swa-12-96000.7z
12067786 leelaz-model-swa-12-96000.bsc
276757452 leelaz-model-swa-12-96000.uncompressed.txt
12432145 leelaz-model-swa-12-96000.ppmd```


The numbers are in bytes.

**Comment by l1t1** [comment]
some more compress by deleting points
eg
4.82e-5 ->482e-7
.000639->639e-6

**Comment by jonlave** [comment]
When are we switching to 40b? There's been several cycles now. Did we face any issues recently that's delaying the switchover?

**Comment by roy7** [comment]
Only thing left is gcp is travelling, once he is back we might make the move. :)

**Comment by l1t1** [comment]
delete point 
```
def format_n(x):
    x = float(x)
    x = '{:.3g}'.format(x)
    x = x.replace('e-0', 'e-')
    if x.startswith('0.'):
        x = x[1:]
    if x.startswith('-0.'):
        x = '-' + x[2:]
    # delete point
    y=x.find('e')
    if y>0:
     z=x.find('.')
     s=x[:y].replace('.','')+'e'+str(int(x[y+1:])-(y-z)+1)
     if len(s)<len(x):return s
    
    return x
```

**Comment by l1t1** [comment]
anyway, the bsc tool of https://github.com/gcp/leela-zero/issues/1733#issuecomment-415744763
is considerable, it's about 0.7 of gz format size.

**Comment by l1t1** [comment]
I also test google brotli, it's performance is much worse

**Comment by l1t1** [comment]
https://github.com/moinakg/pcompress
can use bsc algorithm at linux

**Comment by gcp** [comment]
Note that if we use another codec instead of zlib, we still need to be able to include it in Leela Zero (so it needs to have suitable licensing).

zpf sounds neat. I'll close this as we now have a simple text based quantizer in use.
